sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java

```diff
@@ -0,0 +1,357 @@
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.Symbols;
+import io.crate.lucene.FieldTypeLookup;
+import io.crate.lucene.LuceneQueryBuilder;
+import io.crate.metadata.Reference;
+import io.crate.metadata.doc.DocSysColumns;
+import io.crate.types.DataTypes;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.Version;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.ObjectArray;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.fielddata.IndexOrdinalsFieldData;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.QueryShardContext;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import static io.crate.breaker.RamAccountingContext.roundUp;
+import static io.crate.concurrent.CompletableFutures.failedFuture;
+import static io.crate.execution.dsl.projection.Projections.shardProjections;
+import static io.crate.execution.engine.collect.LuceneShardCollectorProvider.getCollectorContext;
+
+final class GroupByOptimizedIterator {
+
+    static BatchIterator<Row> tryBuild(IndexShard indexShard,
+                                       LuceneQueryBuilder luceneQueryBuilder,
+                                       FieldTypeLookup fieldTypeLookup,
+                                       BigArrays bigArrays,
+                                       InputFactory inputFactory,
+                                       DocInputFactory docInputFactory,
+                                       RoutedCollectPhase collectPhase,
+                                       CollectTask collectTask) {
+        Collection<? extends Projection> shardProjections = shardProjections(collectPhase.projections());
+        GroupProjection groupProjection = getSingleStringKeyGroupProjection(shardProjections);
+        if (groupProjection == null) return null;
+
+        assert groupProjection.keys().size() == 1 : "Must have 1 key if getSingleStringKeyGroupProjection returned a projection";
+        Reference keyRef = getKeyRef(collectPhase.toCollect(), groupProjection.keys().get(0));
+        if (keyRef == null) {
+            return null; // group by on non-reference
+        }
+        MappedFieldType keyFieldType = fieldTypeLookup.get(keyRef.column().fqn());
+        if (keyFieldType == null || !keyFieldType.hasDocValues()) {
```

Comment:
hmm why are the doc values a deal breaker here?

Reply:
The ordinal values wouldn't be available without doc-values.